test-build.ps1

```diff
@@ -0,0 +1,29 @@
+$ErrorActionPreference = 'Stop'
+
+Function Get-Version {
+  Get-ChildItem -directory | Where-Object { ($_.Name -ne "docs") -and ($_.Name -ne "keys") -and ($_.Name -notcontains "chakracore") }
```

Comment:
Match the logic in function.sh

Reply:
@StefanScherer I need help with this function. The equivalent one in bash accepts two arguments one is the path (almost always root) and gets a list of versions, the second one is a filter list that allow limitting the results to only a subset.

This is what allows to use a build matrix.

Reply:
For example you can call `./test-build.sh 6,8 alpine,wheezy` to only build those 4 images.

Reply:
OK, will have a look at it.
My first thought is how I can have a test environment. Sitting at my MacBook with latest Windows Server 1803 I thought I should try this in a container.

```
~/code/docker-node on pull-362
$ uname -a
Darwin Stefans-MacBookPro.local 17.5.0 Darwin Kernel Version 17.5.0: Fri Apr 13 19:32:32 PDT 2018; root:xnu-4570.51.2~1/RELEASE_X86_64 x86_64
~/code/docker-node on pull-362
$ docker run -it -v C:$(pwd):C:$(pwd) -w C:$(pwd) -v //./pipe/docker_engine://./pipe/docker_engine -v "C:/Program Files/docker:C:/docker" microsoft/windowsservercore:1803 powershell
```

Now I have a windowsservercore container running with your PR clone on my Mac mounted into a Windows container. As well as the Docker named pipe from the 1803 VM into the container as well as the Docker binaries.

```
Windows PowerShell 
Copyright (C) Microsoft Corporation. All rights reserved.

PS C:\Users\stefan\code\docker-node> /docker/docker version
Client:
 Version:      17.10.0-ee-preview-3
 API version:  1.33
 Go version:   go1.8.4
 Git commit:   1649af8
 Built:        Fri Oct  6 17:52:28 2017
 OS/Arch:      windows/amd64

Server:
 Version:      17.10.0-ee-preview-3
 API version:  1.34 (minimum version 1.15)
 Go version:   go1.8.4
 Git commit:   b8571fd
 Built:        Fri Oct  6 18:01:48 2017
 OS/Arch:      windows/amd64
 Experimental: false
PS C:\Users\stefan\code\docker-node> 
```

Wow, that works. Now I can edit on my Mac and try the PowerShell script inside a Windows container. This gonna be fun.

Reply:
You can do this on a Mac????

Reply:
Yup: https://github.com/powershell/powershell

Reply:
Yes, you need a Windows "Docker Machine". All my Windows container stuff is done from my Mac with https://github.com/StefanScherer/windows-docker-machine
This emulates such a Windows Docker Machine which I can start (with Vagrant) and then use it with docker-machine env. My Mac home folder is shared with the Windows VM, so volume mounts just look transparent. I only have to prepend "C:" before the Mac "$pwd".

Reply:
```powershell
param ([String[]] $versions, [String[]] $variants)
# Allow comma delimited cli arguments to filter versions and variants
# E.g. .\test-build.ps1 8,10 windowsservercore,nanoserver

$ErrorActionPreference = 'Stop'

Function Get-Version {
  Get-ChildItem -directory | Where-Object { ($_.Name -ne "docs") -and ($_.Name -ne "keys") -and ($_.Name -notcontains "chakracore") }
}

Function Build {
  param ( $version, $variant, $tag )

  $full_tag="$tag-$variant"

  Write-Output "Building node:$full_tag"
  docker build -t node:$full_tag $version/windows/$variant
  if ( $? -eq $FALSE ) {
    Write-Error "Build of $full_tag failed!"
  }
  Write-Output "Testing of $full_tag"
  $path = (Get-Location).path.replace('\', '/')
  docker run --rm -v "$path/:C:/docker-node/" node:"$full_tag" powershell "C:\docker-node\test-image.ps1" "$tag"
}

if (! $versions) {
  $versions = Get-Version
}
if (! $variants) {
  $variants = @('windowsservercore', 'nanoserver')
}

ForEach ($Dir in $versions) {
  $tag = ((Get-Content $Dir\windows\windowsservercore\Dockerfile | Select-String -Pattern 'ENV NODE_VERSION') -split ' ')[2]

  ForEach ($variant in $variants) {
    Build $Dir $variant $tag
  }
}
```

Reply:
Wow, I can even build it when I tag the 1803 images to latest to make the current Windows Dockerfiles happy.

<img width="1386" alt="bildschirmfoto 2018-05-10 um 19 35 11" src="https://user-images.githubusercontent.com/207759/39884216-5ad31536-5489-11e8-8c30-27fbd58d2f61.png">

(the interactive terminal is a bit ugly)

Reply:
Thanks @StefanScherer it works!!